common/util/file_util.cc

```diff
@@ -46,6 +48,46 @@ absl::string_view Stem(absl::string_view filename) {
              : filename.substr(0, last_dot_pos);
 }
 
+// TODO (after bump to c++17) rewrite this function to use std::filesystem
+bool UpwardFileSearch(absl::string_view start, absl::string_view filename,
+                      std::string* result) {
+  static const std::string dir_separator = "/";
+
+  /* Convert to absolute path */
+  char absolute_path[PATH_MAX];
+  if (realpath(start.data(), absolute_path) < 0) {
+    return false;
+  }
+
+  /* Add trailing slash */
+  std::string init = absl::StrCat(absolute_path, dir_separator);
+
+  VLOG(1) << "Upward search for " << filename << ", starting in " << start;
+
+  size_t r = 0, n = init.length();
+  do {
+    r = init.rfind(dir_separator, n);
+
+    if (r == n) {
+      return false;
+    }
+
+    std::string candidate;
+    candidate = absl::StrCat(init.substr(0, r + 1), filename);
+
+    if (access(candidate.c_str(), R_OK) != -1) {
+      *result = candidate;
+      VLOG(1) << "Found: " << *result;
+      return true;
+    }
+
+    n = r - 1;
```

Comment:
With the suggestions above, this now becomes a more readable
`search_up_to = separator_pos - 1;`